saleor/checkout/steps.py

```diff
@@ -16,48 +16,118 @@
+        super(ShippingAddressStep, self).__init__(request, storage, checkout)
+        address_data = storage.get('address', {})
+        self.address = Address(**address_data)
+        self.address_id = storage.get('address_id')
+
         if request.user.is_authenticated():
-            addresses = list(request.user.addresses.all())
-            for address in addresses:
-                data = Address.objects.as_data(address)
-                instance = Address(**data)
-                address.form = self.address_form_class(instance=instance)
-                address.is_selected = Address.objects.are_identical(
-                    address, self.address)
-                if address.is_selected:
-                    existing_selected = True
+            self.email = request.user.email
+            queryset = request.user.addresses.all()
+            addresses = list(queryset)
+            self.addresses = addresses
+
+            if not self.get_same_own_address(self.address):
+                address = self.address
+            else:
+                address = None
+
+            new_address_form = AddressForm(request.POST or None,
+                                           prefix=self.step_name,
+                                           instance=address)
+
+            self.forms['addresses_form'] = UserAddressesForm(
+                data=request.POST or None, queryset=queryset,
+                prefix=self.step_name)
+
+            if not new_address_form.is_bound:
+                selected_address = self.get_same_own_address(self.address)
+                if selected_address:
+                    selected_address.is_selected = True
+                    self.is_new_address = False
+                elif not address_data:
+                    default_address = request.user.default_shipping_address
+                    if default_address:
+                        address = self.get_same_own_address(default_address)
+                        if address:
+                            address.is_selected = True
+                            self.is_new_address = False
+                    elif addresses:
+                        addresses[0].is_selected = True
+                        self.is_new_address = False
         else:
-            addresses = []
-        self.existing_selected = existing_selected
-        self.forms = {'address': address_form}
-        self.addresses = addresses
+            new_address_form = AddressForm(request.POST or None,
+                                           prefix=self.step_name,
+                                           instance=self.address)
+            email = storage.get('email', '')
+            self.email = email
+            self.forms['email'] = AnonymousEmailForm(request.POST or None,
+                                                     initial={'email': email})
+        self.forms['new_address'] = new_address_form
+
+    def get_same_own_address(self, address):
```

Comment:
Maybe call it `get_addressbook_entry`? Another thing: to me `get_*` suggests that not finding a match should result in an exception.